src/coreclr/src/jit/emitarm64.cpp

```diff
@@ -2890,12 +2890,12 @@ emitter::code_t emitter::emitInsCode(instruction ins, insFormat fmt)
  *         'size' specifies the size of the result (16 or 32 bits)
  */
 
-/*static*/ INT32 emitter::emitDecodeByteShiftedImm(const emitter::byteShiftedImm bsImm, emitAttr size)
+/*static*/ INT64 emitter::emitDecodeByteShiftedImm(const emitter::byteShiftedImm bsImm, emitAttr size)
 {
     bool     onesShift = (bsImm.immOnes == 1);
     unsigned bySh      = bsImm.immBY;         // Num Bytes to shift 0,1,2,3
-    INT32    val       = (INT32)bsImm.immVal; // 8-bit immediate
-    INT32    result    = val;
+    INT64    val       = (INT64)bsImm.immVal; // 8-bit immediate
+    INT64    result    = val;
```

Comment:
We first try to encode the immediate using `movi` and if that doesn't work, we try to encode with `movni` by doing the 1s complement of the immediate value. With max values cases, we overflow while trying to do `result <<= (8 * bySh);` since `result` was `INT32`. Making it `INT64` doesn't overflow and set the right value inside `result`. We then compare this value with `imm` inside `canEncodeByteShiftedImm` which itself is `INT64`. We can easily repro it with `int.MaxValue`, `int.MaxValue / 2` , `int.MaxValue / 4`, etc immediates. 

> Can we simplify to?

Sure. Also, I forgot to ask you offline, is there a good place where I can add these test cases?

Reply:
I looked more carefully at emitDecodeByteShiftedImm and still don't think the issue in this function. 

result <<= (8 * bySh) should never overflow since incoming bsImm.immVal is 8 bit value and bySh is less than 4 meaning we left shift 24 bit at maximum (i.e. 7th bit will be at position 31). 

In you example (i.e for int.MaxValue) we should have bySh=3, immVal=0x7F and bsImm.immOnes=1 then result variable after left shift will be 0x‭7F000000‬.

Then we set one to all bits that are below (8 * bySh)-th position by doing 

result |= ((1 << (8 * bySh)) - 1) 

In this case it going to do result |= ‭0xFFFFFF
 
and the final value of result is going to be 0x‭7FFFFFFF.

There is probably some problem with emitter::canEncodeByteShiftedImm where this function is called.

Let's talk offline tomorrow.

Reply:
In the case of `int.MaxValue` we first try to encode in `movi`, and since that doesn't work, we end up passing `immVal=0x80` and `bySh=3`,  `bsImm.immOnes=0` to try `mvni`. With `INT32` we overflow when we do `immVal << 24` and get `-2147483648` or `0xffffffff80000000` while with INT64 we get back `0x80000000` (which was originally 1s complement of `int.MaxValue` (`0x7fffffff`).
By the way, looking at gcc code, looks like that is the right behavior.
https://godbolt.org/z/ytUTzP

But yes, lets talk offline. May be I missed something obvious or may be the assert check should take into account that if it is checking the original value or the 1s complement of it.

Reply:
Discussed offline. The right return type should be `UINT32`.